Feature/openqxd

chaoos opened the pull request:
* Adds interface openQxD <--> QUDA
* Adds capability for C* boundaries in 1,2 or 3 spatial directions
* Fixed CPU reordering
* Adds openQCD gamma matrix basis (and basis transformations)
* Adds order class for spinors
* Adds order class for clover field (U(1) clover field is transfered from openQxD)
* Adds order class for gauge field
* Changed all the comments in quda.h to C89 style comments
* Rebased on your current `develop` branch

mathiaswagner committed:
initial structure (completely untested and still a lot  of things WIP)

mathiaswagner committed:
compilation fixes and plaquette dummy

fernandezdlg committed:
Updated rank formula

The rank formula was updated to the openQxD layout: rank = coords[0]*dims[1]*dims[2]*dims[3] + coords[1]*dims[2]*dims[3]  + coords[2]*dims[3] + coords[3]

fernandezdlg committed:
Typo corrected

The forloop was corrected (typo in i counter)

fernandezdlg committed:
Typo corrected

corrected to i<=3

nyholme committed:
fix rankFromCoords to consider BLK order

mathiaswagner committed:
some fixes for openQCD testing

fernandezdlg committed:
Still not working,

trying to manually couple indexing correctly

fernandezdlg committed:
Now working, todo: add ipt directly in quda

Currently, the reordering is done inside OpenQxd

fernandezdlg committed:
TODO: comments & planning

fernandezdlg committed:
hardwire quda rankfromcoords

fernandezdlg committed:
hardwire test 2

fernandezdlg committed:
Merge branch 'feature/openqxd' of github.com:fernandezdlg/quda into feature/openqxd

fernandezdlg committed:
ignore build folder

fernandezdlg committed:
rankFromCoords for OpenQxD gauge fields working.

Probably it will continue to work for spinor fields... To be confirmed..

fernandezdlg committed:
OpenQCD spinor interface from TIFR (need adjusting

fernandezdlg committed:
commented out openqcd spinors sections for compile

fernandezdlg committed:
spinors feature OK (still needs correct indexing)

fernandezdlg committed:
openqcd dslash correction + load color spinor

fernandezdlg committed:
load-save in spinors+gauge + format

fernandezdlg committed:
small correction on load fct

fernandezdlg committed:
ordering done, lexicographical io w openqxd

fernandezdlg committed:
added openqcd dirac order

fernandezdlg committed:
dslash quda := NO dslash operator applied

fernandezdlg committed:
strange bad indexing behaviour (check3 log)

fernandezdlg committed:
small corrections + comments

In check3, only half the sites are loaded onto quda (odd ones on whole lattice), then in the save function, only the odd ones in the first half get reloaded!

fernandezdlg committed:
cleanup and test oddeven

fernandezdlg committed:
latest test

fernandezdlg committed:
formatting

fernandezdlg committed:
corrected over-include enum + cmmnts

fernandezdlg committed:
return to previous formatting

fernandezdlg committed:
small corrs + formatting

fernandezdlg committed:
gauge save load

fernandezdlg committed:
quda openqcd gauge loadsave, load, and save

fernandezdlg committed:
removed comments

fernandezdlg committed:
rearrangement

fernandezdlg committed:
plaquette only function

fernandezdlg committed:
small correction

fernandezdlg committed:
gauge working, color using spacespinorcolor totest

fernandezdlg committed:
final

chaoos committed:
stated to clean up mess a little

chaoos committed:
cleaned up

chaoos committed:
refactored plaquette

chaoos committed:
Only have generic lexicographical ordering in OpenQCDOrder

chaoos committed:
norms on both sides coincide

chaoos committed:
all reordering of spinor fields done in quda now

chaoos committed:
added functions gamma{0,2,3}() that don't work

Commit:
Add QUDA_OPENQCD_GAMMA_BASIS

chaoos committed:
fixed openQCD_qudaGamma

chaoos committed:
removed implicit fallthrough in switch case of gammai()

chaoos committed:
global minus for openqcd gamma matrix transformation

chaoos committed:
dirac operator seems to work now

chaoos committed:
clover term works

chaoos committed:
GCR inverter works

chaoos committed:
cleaned up a little

donion74 committed:
Add fvtx for profiler

donion74 committed:
Optimize malloc

Commit:
Add parameters for multigrid (incomplete)

Commit:
Undo minus sign in back transformation

donion74 committed:
Add global minus in Dirac

donion74 committed:
Add back and forth test

donion74 committed:
Remove *= -1

Commit:
Revert "Undo minus sign in back transformation"

This reverts commit 271f8285dc6972ae7c8a58e16f8b466c2176493d.

donion74 committed:
Fix back_and_forth test

Commit:
Merge branch 'feature/openqxd-thesis-release' of https://github.com/chaoos/quda into feature/openqxd-thesis-release

donion74 committed:
Fix mistake in transformation and remove Romans' minus sign in transformation

donion74 committed:
Fix mistake in Udagger

Commit:
Modify gamma5

gamma5_openqcd = -1 * U * gamma5_ukqcd * U^dagger

timrupertharris committed:
Changes for MG support.

timrupertharris committed:
Merge branch 'feature/openqxd-thesis-release' of https://github.com/chaoos/quda into multgrid-solver

donion74 committed:
Debug cstar error

donion74 committed:
Debug cstar error

Commit:
Add debugging information

donion74 committed:
Polish code

donion74 committed:
Change back and forth test

donion74 committed:
Merge branch 'multgrid-solver' into feature/openqxd-thesis-release

donion74 committed:
Modify back and forth test

donion74 committed:
Modify printfQuda

donion74 committed:
Merge branch 'multgrid-solver' into feature/openqxd-thesis-release

donion74 committed:
Return true_res and disable half precision in multigrid

donion74 committed:
Remove qudaInvert

donion74 committed:
Use layout.cstar for setting cstar

donion74 committed:
Fix }

Commit:
Merge branch 'feature/openqxd-thesis-release' of https://github.com/chaoos/quda into feature/openqxd-thesis-release

donion74 committed:
Add comment about Precision for multigrid

timrupertharris committed:
Moved init of cuda_prec_sloppy into openQCD_qudaMultigrid.

chaoos committed:
added clover field load function to load cloer field from openQCD

chaoos committed:
we need csw and coeff to determine kappa in clover_field_order.h OpenQCDOrder

chaoos committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/openqxd-thesis-release' into feature/openqxd-thesis-release

chaoos committed:
first attempt for no loads

Commit:
Print debug

Commit:
Use global_ipr for ranksFromCoords

Commit:
Remove debug info

Commit:
Fix bracket, clean up

Commit:
Merge branch 'feature/openqxd-thesis-release' of https://github.com/chaoos/quda into feature/openqxd-thesis-release

Commit:
Comment debugging

Commit:
Remove comment

Commit:
Restore info about communicator

Commit:
Rename variables

Commit:
Remove debug statements

chaoos committed:
added relevant parameters to Dirac struct

Commit:
Use BUILD_INTERFACE_OPENQCD to set cstar

Commit:
Merge branch 'feature/openqxd-thesis-release' of https://github.com/chaoos/quda into feature/openqxd-thesis-release

Commit:
Remove abs(dims)

chaoos committed:
towards a general solver interface

chaoos committed:
works with QCD+QED (clover term is transfered always)

chaoos committed:
finally transfer clover only when needed (ie QCD+QED)

chaoos committed:
added remaining settings

chaoos committed:
solved a bug that prevented reordering on CPU to work

chaoos committed:
in case of openQCD gauge fields, allocate and transfer 4*VOLUME +
7*BNDRY/4 su3 matrices

chaoos committed:
reordering of gauge field done purely within QUDA (only load, not store)

chaoos committed:
changed gauge loading and added eigensolver interface (not working yet)

chaoos committed:
solved the mystical local lattice direction = 6 bug

chaoos committed:
changed comments to C89 style (such that we can link C89 programs
against it)

chaoos committed:
setup verbosity

chaoos committed:
Merge pull request #1 from lattice/develop

Merging upstream develop

chaoos committed:
fixed eigensolverDestroy

chaoos committed:
Merge remote-tracking branch 'origin/develop' into feature/openqxd

chaoos committed:
cleaned up tabs/spaces, comment, unnecessary additions

chaoos committed:
fixed gitignore

chaoos committed:
removed comment

chaoos committed:
cleaned up openqcd interface

chaoos committed:
outsourced all openqcd index functions

chaoos committed:
fixed reordering on GPU

chaoos commented on lib/color_spinor_field.cpp:
```diff
@@ -454,7 +454,7 @@ namespace quda
       if (reorder_location() == QUDA_CPU_FIELD_LOCATION) { // reorder on the host
         void *buffer = pool_pinned_malloc(bytes);
-        qudaMemcpy(buffer, v.data(), bytes, qudaMemcpyDefault);
+        qudaMemcpy(buffer, src.v.data(), bytes, qudaMemcpyDefault);
```
I observed that reordering on the CPU always results in vectors that are zero everywhere. I think there is a typo, since this fixed reordering for me.

maddyscientist replied:
Thanks for catching that.  The CPU ordering path isn't really used much, except for debugging.  I likely introduced this bug in #1384.

chaoos commented on include/clover_field_order.h:
```diff
@@ -11,6 +11,7 @@
 #include <convert.h>
 #include <clover_field.h>
 #include <complex_quda.h>
+#include <index_helper.cuh>
```
I need getCoords as well as my openqcd index functions in the order class

chaoos commented on include/communicator_quda.h:
```diff
@@ -41,6 +41,7 @@ namespace quda
     int (*coords)[QUDA_MAX_DIM];
     int my_rank;
     int my_coords[QUDA_MAX_DIM];
+    int cstar; // number of C* direction as per openQxD convention
```
We extended the Topology struct with a cstar member to indicate how many spatial directions have C* boundaries. Maybe you want to have that as a option in the `QudaInvertParam` struct? However, we need this setting here, since it influences in the process grid topology.

maddyscientist replied:
Forgive my ignorance, but are the C* boundaries only relevant for fermions?

chaoos commented on include/communicator_quda.h:
```diff
@@ -126,9 +127,26 @@ namespace quda
-
-    for (int i = 0; i < QUDA_MAX_DIM; i++) {
-      coords[i] = (i < topo->ndim) ? mod(comm_coords(topo)[i] + displacement[i], comm_dims(topo)[i]) : 0;
+    int shift_integer;
+
+    int Nx_displacement = 0;
+    for (int i = QUDA_MAX_DIM-1; i >=0; i--) {
+      // cstar shift[x] shift[y] shift[z] shift[t]
+      // 0     0        0        0        0
+      // 1     0        0        0        0
+      // 2     0        1        0        0
+      // 3     0        1        1        0
+      if(i < topo->ndim && (
+          (i==1 && topo->cstar >= 2) ||
+          (i==2 && topo->cstar >= 3)
+      )) {
+        // if we go over the boundary and have a shifted boundary condition,
+        // we shift Nx/2 ranks in x-direction:
+        // shift_integer in {-1, 0, 1}
+        shift_integer = (comm_coords(topo)[i] + displacement[i] + comm_dims(topo)[i]) / comm_dims(topo)[i];
+        Nx_displacement += (shift_integer - 1) * (comm_dims(topo)[0]/2);
+      }
+      coords[i] = (i < topo->ndim) ? mod(comm_coords(topo)[i] + displacement[i] + (i==0 ? Nx_displacement :0), comm_dims(topo)[i]) : 0;
```
Here we change the rank-topology based on how many C* directions we have. If there is a C* direction, we calculate the rank displacement in X-direction (`NX_displacement`) accordingly. `topo->cstar=0` recovers legacy behavior.

chaoos commented on include/dslash_quda.h:
```diff
@@ -808,4 +808,10 @@ namespace quda
   */
   void gamma5(ColorSpinorField &out, const ColorSpinorField &in);
 
+  /* RG: I have added these */
+  void gamma0(ColorSpinorField &out, const ColorSpinorField &in);
+  void gamma1(ColorSpinorField &out, const ColorSpinorField &in);
+  void gamma2(ColorSpinorField &out, const ColorSpinorField &in);
+  void gamma3(ColorSpinorField &out, const ColorSpinorField &in);
+
```
I needed gamma0 to gamma3 for tests in openQxD side. Let me know if there is an intended way to apply a gamma matrix other than gamm5 to a spinor field (I didn't find anything).

maddyscientist replied:
I don't think we've needed anything other in gamma5 in the past, so I never bothered to instantiate them.  No problem adding them and this is the correct place to do so.

I do question the difference in notation here though, for C-style counting from zero for gamma_1..gamma_4 and Fortran-style / physics notation for gamma_5.  Should we make these consistent?

weinbe2 replied:
I think a better solution could be creating `enum` for `QUDA_GAMMA_X`, `QUDA_GAMMA_Y`, `QUDA_GAMMA_Z`, `QUDA_GAMMA_T`, and `QUDA_GAMMA_5` (or something of that sort), then exposing a version of `ApplyGamma` that takes in these arguments. This can then be a thin wrapper around the `ApplyGamma` function that takes in 0 ... 4, and the thin wrapper can handle mapping the `enum` to the appropriate integers internally.

chaoos replied:
Good point, I removed the functions `gamma{0,1,2,3}` and wrote a `ApplyGamma` that takes an enum as direction, see https://github.com/lattice/quda/pull/1414/commits/d7fe0a237022b7f6cda97c8624a7d723ae39a5bb and https://github.com/lattice/quda/pull/1414/commits/bce45981bdbec211f522a2fc2f21a8c95ab88a2f

chaoos commented on include/enum_quda.h:
```diff
@@ -18,18 +18,18 @@ typedef enum QudaMemoryType_s {
-//
-// Types used in QudaGaugeParam
-//
+/**
```
Also here I changed all the comments to C89 style ones to be able to compile openQxD in C89. People don't want to change that standard ...

maddyscientist replied:
That's unfortunate, and will be a difficult constraint going forward as we have been planning on making the interface fully embrace C99 (e.g., use of `bool`).

What's the reason for this desire to stick to a 35 year old standard?

chaoos commented on include/gauge_field_order.h:
```diff
@@ -23,6 +23,7 @@
 #include <load_store.h>
 #include <aos.h>
 #include <transform_reduce.h>
+#include <communicator_quda.h>
```
I need the number of ranks in all 4 directions (see `nproc` in the order class) to calculate offsets for the gauge, clover and spinor fields. Let me know if there is a better way to do that. in openQCD these are compile time constants.

chaoos replied:
This line is removed.

chaoos commented on include/index_helper.cuh:
```diff
@@ -1106,4 +1106,302 @@ namespace quda {
     return (((x[3]*X[2] + x[2])*X[1] + x[1])*X[0] + x[0]) >> 1;
   }
 
+
```
Here in a namespace `openqcd`, I added all indexing helper function that are needed by the order classes to reduce code duplication.

chaoos commented on include/quda.h:
```diff
@@ -16,7 +16,7 @@
 #ifndef __CUDACC_RTC__
 #define double_complex double _Complex
-#else // keep NVRTC happy since it can't handle C types
+#else /* keep NVRTC happy since it can't handle C types */
```
Again the C89 style comments

chaoos commented on include/quda.h:
```diff
@@ -980,7 +979,7 @@ extern "C" {
    * previously called initQudaDevice.
    */
-  void initQudaMemory();
+  void initQudaMemory(void);
```
C89 complains that without the void, this is not a valid function signature.

weinbe2 replied:
Just double-checking, is this still a concern, or has OpenQxD moved to C99?

chaoos replied:
Unfortunately yes. Also in C99 with strict-prototypes this will error. I would prefer to leave it if that's no problem. But I added the void also in the function definition (I missed doing that). Now it's consistent at least. This way it is C-idiomatic (although probably not C++-idiomatic).

weinbe2 replied:
I'll need to make sure it doesn't break MILC for some strange reason (I think it'll be fine), but otherwise it is what it is and I accept it!

maddyscientist replied:
Just wondering if we could augment the C99 compilation test (`tests/c_interface_test.c`) to enable strict prototypes to ensure this isn't accidentally regressed?

weinbe2 replied:
MILC compiled just fine!

chaoos commented on lib/comm_common.cpp:
```diff
@@ -102,6 +102,13 @@ namespace quda
     Topology *topo = new Topology;
 
+#ifdef BUILD_OPENQCD_INTERFACE
+    int *data = static_cast<int *>(map_data);
+    topo->cstar = data[0];
+#else
+    topo->cstar = 0;
+#endif
+
```
Here we set `cstar` to the value that comes from openQxD. We set `cstar=0` else, to recover legacy behavior.

maddyscientist replied:
To avoid possible issues in the future, best to make sure that any local variables in each path here are scoped or simply write the first path here as follows to avoid any local variable:
```c++
topo->cstar = static_cast<int *>(map_data)[0];
```

chaoos commented on lib/dslash_gamma_helper.cu:
```diff
@@ -103,4 +103,9 @@ namespace quda {
   // Applies a gamma5 matrix to a spinor (wrapper to ApplyGamma)
   void gamma5(ColorSpinorField &out, const ColorSpinorField &in) { ApplyGamma(out,in,4); }
 
+  /* RG: I have added these */
+  void gamma0(ColorSpinorField &out, const ColorSpinorField &in) { ApplyGamma(out,in,0); }
+  void gamma1(ColorSpinorField &out, const ColorSpinorField &in) { ApplyGamma(out,in,1); }
+  void gamma2(ColorSpinorField &out, const ColorSpinorField &in) { ApplyGamma(out,in,2); }
+  void gamma3(ColorSpinorField &out, const ColorSpinorField &in) { ApplyGamma(out,in,3); }
```
Again, let me know it I should remove them.

chaoos commented on lib/gauge_field.cpp:
```diff
@@ -151,6 +151,25 @@ namespace quda {
         bytes = site_dim * (x[0] + 4) * (x[1] + 2) * (x[2] + 2) * (x[3] + 2) * nInternal * precision;
       } else if (order == QUDA_MILC_SITE_GAUGE_ORDER) {
         bytes = volume * site_size;
+      } else if (order == QUDA_OPENQCD_GAUGE_ORDER) {
+        /**
+         * With an openQCD gauge field, we need all the links of even lattice
+         * points in positive direction. These are links that lie in the buffer
+         * space that spans 7*BNDRY/4 gauge fields. These boundary fields are
+         * located at base_ptr + 4*VOLUME. Therefore, we need to transfer more
+         * than 4*VOLUME matrices.
+         */
+
+        /* analogue to BNDRY in openQCD:include/global.h */
+        long int bndry = 0;
+        bndry += (1-(comm_dim(0)%2))*x[1]*x[2]*x[3];
+        bndry += (1-(comm_dim(1)%2))*x[0]*x[2]*x[3];
+        bndry += (1-(comm_dim(2)%2))*x[0]*x[1]*x[3];
+        bndry += (1-(comm_dim(3)%2))*x[0]*x[1]*x[2];
+        bndry *= 2;
+
+        length += 18*7*bndry/4;
+        bytes = length * precision;
```
For openQxD, we need to copy a little more data for the gauge field, because the gauge fields of even boundary points in positive direction are stored in `base_pointer + 4*VOLUME + some_offset`

donion74 commented on include/clover_field.h:
```diff
@@ -133,6 +133,8 @@ namespace quda {
       clover(param.clover),
       cloverInv(param.cloverInv),
+      csw(param.csw),
+      coeff(param.coeff),
```
What does this `coeff` do?

chaoos replied:
This is set to `kappa*csw`. in `loadCloverQuda()` and later used in `computeClover()`.

donion74 replied:
Do you want to add this comment in the code?

donion74 commented on include/index_helper.cuh:
```diff
@@ -1106,4 +1106,302 @@ namespace quda {
+    }
+
+
+    /**
```
Could you write out what `cbs`does?

chaoos replied:
Comment added

donion74 commented on include/index_helper.cuh:
```diff
@@ -1106,4 +1106,302 @@ namespace quda {
+    }
+
+
+    __device__ __host__ inline int ifc(const int *L, const int *nproc, const int mu)
```
Could you add a small comment what this does?

chaoos replied:
Comment added

donion74 commented on include/index_helper.cuh:
```diff
@@ -1106,4 +1106,302 @@ namespace quda {
+
+
+    /**
+     * @brief      Pure implementation of iup[ix][mu]
```
Maybe add: Returns neighbouring point in positive $\mu$ direction.

chaoos replied:
Comment added

chaoos committed:
macro war to enable "-std=c89 -pedantic -Werror" in openqxds checks

chaoos commented on include/quda_openqcd_interface.h:
```diff
@@ -0,0 +1,343 @@
+/**
+ * The macro battle below is to trick quda.h to think that double_complex is
+ * defined to be the struct below. For this we need to set the __CUDACC_RTC__,
+ * which makes double_complex to be defined as double2 (see quda.h), which we
+ * redefine below as openqcd_complex_dble. The original definitions of
+ * __CUDACC_RTC__ and double2 are recovered below. We do this to be able to
+ * include this header file into a openQxD program and compile with flags
+ * "-std=C89 -pedantic -Werror". Else the compiler trows an
+ * "ISO C90 does not support complex types" error because of the
+ * "double _Complex" data types exposed in quda.h.
+ */
+
+typedef struct
+{
+  double re,im;
+} openqcd_complex_dble;
+
+#ifdef __CUDACC_RTC__
+#define __CUDACC_RTC_ORIGINAL__ __CUDACC_RTC__
+#endif
+
+#ifdef double2
+#define double2_ORIGINAL double2
+#endif
+
+#define __CUDACC_RTC__
+#define double2 openqcd_complex_dble
+#include <quda.h>
+#undef double2
+#undef __CUDACC_RTC__
+
+#ifdef double2_ORIGINAL
+#define double2 double2_ORIGINAL
+#undef double2_ORIGINAL
+#endif
+
+#ifdef __CUDACC_RTC_ORIGINAL__
+#define __CUDACC_RTC__ __CUDACC_RTC_ORIGINAL__
+#undef __CUDACC_RTC_ORIGINAL__
+#endif
```
I know that looks ugly, but it redefines the datatype `double2` to the struct above, such that in `quda.h` the `double_complex` is redefined to `double2` (which in turn is `openqcd_complex_dble`), which is something that C89 is not complaining about. I agree that we should probably solve that differently :-), but by this we can still build on the openqxd side with `-std=c89 -pedantic -Werror`.

maddyscientist replied:
This makes me wonder what hoops might be necessary when we switch to using `bool` in the interface...

chaoos committed:
added comment for loadGaugeQuda

chaoos committed:
added MatQuda profiling

maddyscientist commented:
@Jenkins test this please

maddyscientist commented:
Thanks for this contribution @chaoos.  This looks like a great PR and it will be great to OpenQCD support plumbed into QUDA 😄 

@mathiaswagner and I are traveling at the moment, so reviewing this will likely take longer than normal.  In the meantime, can you fix the broken builds in the CI matrix? (click on the `Details` link next to the failed build, and you should be able to find the reason for the failure).  Also, please add your name to the `README` under the list of contributors.

havogt commented:
cscs-ci run

chaoos committed:
removed check functions and unnecessary including

chaoos committed:
added requested comments

chaoos commented:
@maddyscientist Could you trigger the pipeline again? Locally I fixed the problems.
Thank you,
Roman

mathiaswagner commented:
@Jenkins test this please.

maddyscientist commented on include/clover_field_order.h:
```diff
@@ -1015,6 +1016,102 @@ namespace quda {
+      /**
+       * OpenQCD ordering for clover fields
+       */
+      template <typename Float, int length = 72> struct OpenQCDOrder {
```
Just curious, whether QUDA has sufficient capability to create the clover term for OpenQ*D's use case or whether you have specific requirements that we are lacking?  E.g., exponential clover or anisotropy?

weinbe2 replied:
@maddyscientist filling this in for future reference, this is relevant for QCD+QED simulations where the QCD and QED clovers need to be separately constructed from the SU(3) and U(1) fields, respectively, then combined. We (QUDA) would need to be able to explicitly handle standalone U(1) fields.

The reason we can't do this directly from the unified U(3) = SU(3)xU(1) fields is because you can't (naively) disentangle the SU(3) and U(1) components. (...I guess you can divide the U(1) contribution out of the SU(3) field if you know the right relative scaling coefficient, but it would be pretty gross)

maddyscientist commented on include/color_spinor_field_order.h:
```diff
@@ -1811,6 +1811,90 @@ namespace quda
+        dim {a.X(0), a.X(1), a.X(2), a.X(3)}, // *local* lattice dimensions, xyzt
+        L   {a.X(3), a.X(0), a.X(1), a.X(2)}  // *local* lattice dimensions, txyz
+      {
+        if constexpr (length != 24) {
```
Make this a `static_assert`?

maddyscientist commented on include/gauge_field_order.h:
```diff
@@ -2347,8 +2347,124 @@ namespace quda {
+        L   {u.X()[3], u.X()[0], u.X()[1], u.X()[2]}, // *local* lattice dimensions, txyz
+        nproc {comm_dim(3), comm_dim(0), comm_dim(1), comm_dim(2)} // txyz
+      {
+        if constexpr (length != 18) {
```
`static_assert` here?

maddyscientist commented on include/index_helper.cuh:
```diff
@@ -1106,4 +1106,323 @@ namespace quda {
+
+      if ((x[mu]==(L[mu]-1))&&(nproc[mu]>1)) {
+
+        int cbs[4] = {setup_cbs(0, L), setup_cbs(1, L), setup_cbs(2, L), setup_cbs(3, L)};
```
Might not be practical, but in an ideal world the `cbs` numbers would be pre-computed on the host in `int_fastdiv` types.  This would allow for much faster division in the kernels.  Likely not performance critical, just mentioning it out of completeness.

maddyscientist commented on lib/copy_gauge_extended.cu:
```diff
@@ -110,6 +110,15 @@ namespace quda {
+      using G = OpenQCDOrder<FloatOut, length>;
+      CopyGaugeEx<FloatOut, FloatIn, length, G, InOrder>(out, in, location, Out, In);
+#else
+      errorQuda("OPENQCD interface has not been built\n");
```
General comment: `errorQuda` includes a line break (`\n`), so these should be removed.

maddyscientist commented on lib/gauge_field.cpp:
```diff
@@ -151,6 +151,25 @@ namespace quda {
+        /**
+         * With an openQCD gauge field, we need all the links of even lattice
+         * points in positive direction. These are links that lie in the buffer
+         * space that spans 7*BNDRY/4 gauge fields. These boundary fields are
```
Curious where the factor of 7 comes from?

maddyscientist commented on lib/openqcd_interface.cpp:
```diff
@@ -0,0 +1,1360 @@
+  {"QUDA_CUSOLVE_EXTLIB",                   std::to_string(QUDA_CUSOLVE_EXTLIB)},
+  {"QUDA_EIGEN_EXTLIB",                     std::to_string(QUDA_EIGEN_EXTLIB)},
+  {"QUDA_EXTLIB_INVALID",                   std::to_string(QUDA_EXTLIB_INVALID)}
+};
```
This is something we should probably standardize this for all enums and make available library wide.  @mathiaswagner did I imagine us talking about something like this a few years ago?

mathiaswagner replied:
Yes, we had discussed that but seems we lost track.

maddyscientist reviewed (commented):
Thank you for this PR, this is a great contribution to QUDA.  I have done an initial visual review at this point, and left a bunch of comments.  My only real concern of note is the demotion of the interface to being C89 instead of C99, but those concerns may be unwarranted.

Outside of those comments, somethings that will need to be done:

- [ ] Apply clang-format (see [here](https://github.com/lattice/quda/wiki/Coding-Conventions-and-Style#coding-style-and-clang-format))
- [ ] Create a wiki page of how to use OpenQ*D with QUDA, similar for example to what we have for [MILC}(https://github.com/lattice/quda/wiki/MILC-with-QUDA)

Is there any specific functionality that is added to QUDA that we should be adding to the internal tests?

chaoos committed:
check for nullptr in comm_create_topology

chaoos committed:
ipt and iup function exposed, parsing infile section "QUDA"

chaoos committed:
the Wilson and Wilson-clover Dirac operator are gamma5 Hermtitian

chaoos committed:
added functionality to compute both (left and right) singular vectors in computeSVD

chaoos committed:
added comments for eigensolver interface

chaoos committed:
some convenience mappings for the enums

chaoos committed:
in TRLM: n_kr must be >= n_conv+12, else iter_keep may become negatice
(ends in a segfault)

chaoos committed:
verbose error message if section is not a quda section

chaoos committed:
more verbose error msg when solver section is not a QUDA one

chaoos committed:
added function comm_dim_cstar to query if a dimension is C*

chaoos committed:
solved P2P communication bug when C* boundaries are active

chaoos committed:
added packed formats

chaoos committed:
Merge branch 'feature/openqxd' of https://github.com/chaoos/quda into feature/openqxd

chaoos committed:
dynamic t_boundary

chaoos committed:
enable RECONSTRUCT_8/9/12/13/NO for openqxd interface

chaoos committed:
fixed eigensolver, when cuda_prec_eigensolver is not set in the infile

chaoos committed:
only transfer/recalc gauge/clover field according to openqxd's flag DB

chaoos committed:
added void pointer to QudaInvertParam struct for arbitrary additinal
properties

chaoos committed:
changed solver interface to only deal with solver section IDs

chaoos committed:
refactored interface

chaoos committed:
removed ptr in getHandle

chaoos committed:
removed ptr in getHandle (2)

chaoos committed:
fixed clang complaining "error: suggest braces around initialization of
subobject"

chaoos committed:
in openQCD sometimes we test with gauge field all set to unity, then the
Sw term is zero, but QUDA should not complain then

chaoos committed:
remove trailing whitespaces

chaoos committed:
applied clang-format as described in https://github.com/lattice/quda/wiki/Coding-Conventions-and-Style#coding-style-and-clang-format

chaoos commented:
> Thank you for this PR, this is a great contribution to QUDA. I have done an initial visual review at this point, and left a bunch of comments. My only real concern of note is the demotion of the interface to being C89 instead of C99, but those concerns may be unwarranted.
> 
> Outside of those comments, somethings that will need to be done:
> 
>     * [ ]  Apply clang-format (see [here](https://github.com/lattice/quda/wiki/Coding-Conventions-and-Style#coding-style-and-clang-format))
> 
>     * [ ]  Create a wiki page of how to use OpenQ*D with QUDA, similar for example to what we have for [MILC}(https://github.com/lattice/quda/wiki/MILC-with-QUDA)
> 
> 
> Is there any specific functionality that is added to QUDA that we should be adding to the internal tests?

Dear Kate,

I dealt with the two points: 

* see commit https://github.com/lattice/quda/pull/1414/commits/568dd3734e4aa20b99e0e18bd1f7d344e5e5bb9a for the clang-formatting
* The page for the wiki is written, but I cannot commit it, permission issue. I attach the patch here [openqxd.zip](https://github.com/lattice/quda/files/14418468/openqxd.zip)

Best,
Roman

maddyscientist commented:
Hi Roman,

I've added you to the lattice github group, so you should have permission to edit the wiki now (and push to branches in the lattice org).

Thx

maddyscientist commented:
@Jenkins ok to test

chaoos committed:
eigensolver iterface rewritten to use identifiers too (as inverter
interface)

chaoos committed:
fixed compiler error in CI/CD

chaoos committed:
clang-format

chaoos commented:
I was able to commit to the wiki, thanks for giving me access!
The CI/CD also seems to be fixed.

havogt commented:
cscs-ci run

chaoos committed:
fixed eigensolver deallocation

maddyscientist commented on include/dirac_quda.h:
```diff
@@ -2606,6 +2606,8 @@ namespace quda {
           && (pc_type == QUDA_MATPC_EVEN_EVEN || pc_type == QUDA_MATPC_EVEN_EVEN_ASYMMETRIC))
         return true;
 
+      if (dirac_type == QUDA_WILSON_DIRAC || dirac_type == QUDA_CLOVER_DIRAC) return true;
```
This addition here looks bogus: the Wilson operator is not Hermitian.  Is this addition intentional?

chaoos replied:
This is in the class `DiracG5M` (i.e. gamma5 times the Dirac operator), which is Hermitian. I had to change this, since the eigensolver didn't accept a non-Hermitian operator else. So this should be correct (only for `gamma5*D` and `D*gamma5` though).

weinbe2 replied:
So this technically isn't proper, because `hermitian()` is for Hermitian positive definite, not just general Hermitian. This check is used for things like the Lanczos eigensolver, which is for HPD only. (If Lanczos is working for Hermitian-indefinite operators, that's actually... insane, and we should talk about it separately @chaoos )

To handle gamma5 ops correctly, you need to go through the Arnoldi solver, which is documented here: https://github.com/lattice/quda/wiki/QUDA%27s-eigensolvers#computing-the-gamma5-hermitian-spectrum

chaoos replied:
I see, removed https://github.com/lattice/quda/pull/1414/commits/49211c40dcea8c1ef21c0910bc70748053520a8b

maddyscientist commented on include/instantiate_dslash.h:
```diff
@@ -38,6 +38,20 @@ namespace quda
 #else
       errorQuda("QUDA_RECONSTRUCT=%d does not enable reconstruct-8/9", QUDA_RECONSTRUCT);
 #endif
+#ifdef BUILD_OPENQCD_INTERFACE
```
@kostrzewa noting that we could augment this for tmLQCD after this is merged to give recon 9/13 support, which I guess will help your perf?

maddyscientist reviewed (changes requested):
I've had a review of the latest code.  This is a great contribution, but a few things left outstanding to be resolved.  I have left some comments, but beyond those:
* The C89 changes are of concern, since we will be embracing a C99 interface, e.g., use of `bool`.
* The new openQCD interface is rather opaque to me.  Can this be better documented?
* Thanks for the wiki additions.  Is there a test input file or equivalent to allow us to run a test with this?

chaoos commented:
> The C89 changes are of concern, since we will be embracing a C99 interface, e.g., use of `bool`.

Dear Kate,
* The C99 thing is ok from our side. Should I just revert all changes I made in quda.h regarding this?
* We are currently in the phase of documenting, so give me some days for this. Do you want more comments in the code (I think every function/method has a javadoc-decoration), or conceptual things in the wiki?
* Regarding testing of openqxd functionality; we have some testing on the side of openqxd. If you want to test this in this repo, then you need to clone and compile the openqxd repo and run the tests from there. We are working on making this repo public, but there is still some politics going on: that's not in my power.

maddyscientist commented:
@chaoos apologies for the slow reply at my end.  GTC and other travel got in the way.....

If you can revert the C99 -> C89 changes that would be great.  It would also narrow the scope of the PR which would also be helpful.

Regarding documentation, the basic rule is that all new functions should have doxygen markup next to their declarations, together with overall conceptual documentation in the wiki.  If you can write up the documentation to include cloning the openq*d and compiling the code to work with QUDA that would be great.  Understood that the repo isn't public yet, and that's fine for now.  Does that sound ok to you?

chaoos committed:
added openQCD_qudaDw_NoLoads (wrapper around Dirac operator that does
not do any transfer of fields)

chaoos committed:
Merge remote-tracking branch 'upstream/develop' into feature/openqxd

chaoos committed:
Merge remote-tracking branch 'upstream/develop' into feature/openqxd

chaoos committed:
remove minimize_memory

chaoos committed:
solves https://github.com/lattice/quda/issues/1532

chaoos committed:
remove C89 comments

chaoos committed:
revert all the c89 changes

Commit:
remove default geo_block_size of 4x4x4x4 for level 0

chaoos committed:
added back newline at EOF

chaoos committed:
Merge remote-tracking branch 'upstream/develop' into feature/openqxd

chaoos committed:
fix when compile with QUDA_DIRAC_DEFAULT_OFF=OFF

chaoos committed:
fix problem, when qhat changes when U(3) gauge group -> need transfer of gauge/clover

chaoos committed:
apply clang-format

chaoos committed:
reverted changes in interface_quda.cpp

chaoos commented:
Hello

I'm sorry for the long wait! Finally I was able to finish this up.

I've reverted all the C89 changes, appled clang-format again, implemented CI/CD testing for all this on our side, and fixed the issues we still had.

So nothing really stops this anymore from being merged. Please review again. Let me know if there is something I can do to settle this.

Thanks and best wishes,
Roman

PS: I've created a spack package for QUDA (pull-request awaits approval, see https://github.com/spack/spack/pull/48939). This became very useful for our CSCS CI/CD workflow. Maybe it's also useful for others, and can be put in the wiki or so?

chaoos committed:
added asych solver call

chaoos committed:
moved structs, added async interface functions

chaoos committed:
fixed async with MPICH

chaoos committed:
nullptr to MPI_COMM_NULL

chaoos committed:
fixed async solver call for cray MPI

chaoos committed:
first implementation of two grids

chaoos committed:
added bcasts for status and residual in invert

chaoos committed:
remove manual reordering from quda to openqcd

chaoos committed:
cleaned up mess

chaoos committed:
WITH_COMM() and in_comm()

chaoos committed:
Merge pull request #5 from chaoos/feature/openqxd-two-grids

Feature/openqxd-two-grids into Feature/openqxd

chaoos committed:
connect to multi RHS interface functions

weinbe2 commented:
@chaoos following up on last week's meeting---if you can resolve the new merge conflicts, we should be able to get this merged this week

chaoos committed:
Merge branch 'develop' into feature/openqxd

weinbe2 commented on include/enum_quda_fortran.h:
```diff
@@ -47,6 +47,7 @@
 #define QUDA_BQCD_GAUGE_ORDER 15 // expect *gauge mu even-odd spacetime+halos row-column order
 #define QUDA_TIFR_GAUGE_ORDER 16
 #define QUDA_TIFR_PADDED_GAUGE_ORDER 17
+#define QUDA_OPENQCD_GAUGE_ORDER 18
```
This (and other `enum`) will need to be updated once `feature/vectorize` is merged

weinbe2 commented on include/quda_openqcd_interface.h:
```diff
@@ -0,0 +1,525 @@
+
+/**
+ * OpenQCD inmitations of structs.
+ * Copied from flags.h, su3.h
```
We can't have code directly copied from `OpenQCD`. However, I see that the data layouts are publicly documented in the `docs` subdirectory of the OpenQCD code, as is noted in the `README` file, which I didn't need to look at the code to see.

Can you remove this explicit copy+paste and implement data structures, such as `openQCD_bc_params_t`, that match the data layout in the documentation, and cite the documentation appropriately?

chaoos replied:
I changed the names and implmementations of these structs + I remove an unneccessary one. They don't have to be alignment compatible anymore and a reference to the docs is added.

weinbe2 commented:
@chaoos are there ever cases where OpenQ*D would need both recon-13/9 _and_ recon-12/8 simultaneously? I'm wondering if we should make QCD+QED fields a compile-time option where, if it's enabled, Wilson-type ops are recon 18/13/9 and otherwise are 18/12/8. Otherwise we might need to go through some plumbing headaches to make sure, for ex, 18/13/12/9/8 are _all_ unit tested in a single build.

@maddyscientist what are your thoughts? I think it's good that we have support for QCD+QED workflows more broadly, but supporting five different reconstructs in a single build is unit-testing nightmare fuel.

weinbe2 commented on lib/openqcd_interface.cpp:
```diff
@@ -0,0 +1,2228 @@
+  setMPICommHandleQuda(&layout.quda_comm);
+
+#ifdef MULTI_GPU
+/* TODO: would we ever want to run with QMP COMMS? */
```
If `OpenQCD`/`OpenQxD` doesn't support QMP comms, then I think it's best to make the build error out if you try to compile with both the OpenQCD interface and QMP comms---if anything because there's no means to test it. More ideally, this check would be done at `cmake` config time.

If some need for OpenQCD + QMP comes up in the future, we can revisit it then.

chaoos replied:
True, I removed the QMP part and added a cmake condition to error if `QUDA_QMP=ON` and `QUDA_INTERFACE_OPENQCD=ON`.

weinbe2 commented on lib/openqcd_interface.cpp:
```diff
@@ -0,0 +1,2228 @@
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define WITH_COMM(expr) do { if (in_comm()) { expr; } } while (0)
```
Two things:

1. You should put `#undef WITH_COMM` at the end of this file, just as a best-practices.
2. Out of curiosity, why does this need to be in `do {...} while (0)`?

chaoos replied:
1. Added, thanks!
2. That one of these C shenanigans one has to be careful. If you have
```
if (x)
  WITH_COMM(foo());
else
  bar();
```
and the macro without the `do { } while (0)` the `else` binds to the inner `if`, not the outer one.

weinbe2 replied:
Wow, okay, thank you for explaining this.

weinbe2 commented on lib/eig_trlm.cpp:
```diff
@@ -26,7 +26,8 @@ namespace quda
     // Thick restart specific checks
-    if (n_kr < n_ev + 6) errorQuda("n_kr=%d must be greater than n_ev+6=%d\n", n_kr, n_ev + 6);
+    if (n_kr < n_ev + 6) errorQuda("n_kr=%d must be greater than or equal to n_ev+6=%d\n", n_kr, n_ev + 6);
+    if (n_kr < n_conv + 12) errorQuda("n_kr=%d must be greater than or equal to n_conv+12=%d\n", n_kr, n_conv + 12);
```
What was the reason for this? Was this a check that was missing all along, or was it empirical?

chaoos replied:
That's long ago and I forgot, but I found my old commit message, see https://github.com/chaoos/quda/commit/ae170fe25f1bd80dec57ae1a39ea5d328f9ec20c
Trusting my past-me, it actually makes sense if yo look at https://github.com/chaoos/quda/blob/feature/openqxd/lib/eig_trlm.cpp#L127

weinbe2 replied:
Oh awesome, good catch!

weinbe2 commented:
@chaoos sorry this stalled for a while, I added some review comments. We can also schedule a call at some point to go over it.

chaoos committed:
Merge remote-tracking branch 'upstream/develop' into feature/openqxd

chaoos committed:
fix segfault in errorQuda

chaoos committed:
anonymous union in struct is not C99

chaoos committed:
added QUDA_QCD_PLUS_QED cmake flag to build Wilson-ops with reconstruct_no/8/12 or no/9/13

chaoos committed:
openqcd interface uses QUDA_QCD_PLUS_QED now

chaoos committed:
remove copy/paste structs from openQCD, but reimplement them / remove spinor_dble struct (not necessary)

chaoos committed:
remove QMP comms from openQCD interface, added cmake error if QUDA_QMP=ON and QUDA_INTERFACE_OPENQCD=ON

chaoos committed:
added #undef WITH_COMM at the end of the file

chaoos commented:
> are there ever cases where OpenQ*D would need both recon-13/9 _and_ recon-12/8 simultaneously?

Originally yes, but I saw that it makes stuff increasingly complicated. So I added a new Cmake flag `QUDA_QCD_PLUS_QED=ON/OFF` as you proposed to build Wilson ops with `QUDA_RECONSTRUCT_NO/8/12` if set to `OFF` (the default) or `QUDA_RECONSTRUCT_NO/9/13` if set to `ON`.

See https://github.com/lattice/quda/pull/1414/commits/452f6e67d8922262ce6c34013220d1efbe156832

chaoos committed:
removed G5D as HPD op

chaoos committed:
removed unneccessary adds

chaoos committed:
exposed ApplyGamma and added enum for direction: QUDA_GAMMA_X, QUDA_GAMMA_Y, QUDA_GAMMA_Z, QUDA_GAMMA_T, QUDA_GAMMA_5

chaoos committed:
add me to the list of authors

chaoos committed:
remove gamma{0,1,2,3}

Commit:
fix error in pipeline, when dual_grid+cray-mpipch+C*-lattice+mrhs+async

weinbe2 commented on include/index_helper.cuh:
```diff
@@ -1109,6 +1109,295 @@ namespace quda {
+     *
+     * @return     Surface
+     */
+    __device__ __host__ inline int surface(const int *X, const int mu)
```
Can we replace `const int *X` with `const int X[4]` to be consistent with other kernels in `include/index_helper.cuh`, or is there a specific reason this needs to be a pointer? (Similar question for other routines below.)

chaoos replied:
Done for all routines https://github.com/lattice/quda/pull/1414/commits/be3ba61cfbd69f71c7e7c384113aa48a5761a49b

weinbe2 commented on include/quda_openqcd_interface.h:
```diff
@@ -0,0 +1,522 @@
+#pragma once
+
+#include "mpi.h"
```
Can these be changed to `<mpi.h>` and `<pthread.h>`, or is this an OpenQxD requirement?

Also, if these are being baked in, these should also be guarded by a check that QUDA was specifically compiled with MPI, since QUDA does still support single GPU/single MPI rank compilation.

chaoos replied:
Changed and added a cmake flag to build openqcd interface always with MPI enabled, see https://github.com/lattice/quda/pull/1414/commits/c7e3f74f60b5876c634887e0ad4b849f67ef58f2

weinbe2 commented on lib/copy_gauge_extended.cu:
```diff
@@ -65,7 +65,7 @@ namespace quda {
         errorQuda("QUDA_RECONSTRUCT=%d does not enable reconstruct-8", QUDA_RECONSTRUCT);
 #endif
-#ifdef GPU_STAGGERED_DIRAC
+#if defined(GPU_STAGGERED_DIRAC) || defined(BUILD_OPENQCD_INTERFACE)
```
`defined(BUILD_OPENQCD_INTERFACE)` should be replaced by `defined(QUDA_QCD_PLUS_QED)` as appropriate

chaoos replied:
True, changed https://github.com/lattice/quda/pull/1414/commits/abbedf7dfa83805b282f86ecd2b53fc8cfde377f, https://github.com/lattice/quda/pull/1414/commits/9b7525d2639c4648843dfabfc7f42f5e01c612d0

weinbe2 commented on lib/copy_gauge_extended.cu:
```diff
@@ -111,10 +111,18 @@ namespace quda {
 
+    } else if (out.Order() == QUDA_OPENQCD_GAUGE_ORDER) {
+
+#ifdef BUILD_OPENQCD_INTERFACE
```
And this is an exception for where we'd need it, of course...

weinbe2 commented on include/quda.h:
```diff
@@ -70,10 +70,8 @@ extern "C" {
-      bool use_split_gauge_bkup; /**< Used by gauge split buffers (default=true keep split gauge after usage)*/
-      int pad;                   /**< Forces 4-byte alignment */
-    };
+    bool use_split_gauge_bkup; /**< Used by gauge split buffers (default=true keep split gauge after usage)*/
```
I did some digging into the C/C++ standard and apparently there's no guarantee that `bool` is 1 byte. In practice, this seems to be the case, but since it's technically implementation-defined I'd like to make this more robust.

Can this be replaced by an `int` and then in various codes it's explicitly compared with 0 to check `true`/`false`? In that case it'd be safer...

chaoos replied:
I changed it to an int and replaced occurences approriately, see https://github.com/lattice/quda/pull/1414/commits/ddc9730a760ee688a30a22211f51af4759549951

chaoos committed:
change pointer to annotated array

chaoos committed:
standard mpi and pthreads including and guard that quda was compiled with MPI if openqcd interface is on

chaoos committed:
build reconstruct-9/13 if QUDA_QCD_PLUS_QED=ON

chaoos committed:
change use_split_gauge_bkup from bool to int

chaoos committed:
reference to code documentation

chaoos committed:
definitions match declarations

chaoos committed:
Merge remote-tracking branch 'upstream/develop' into feature/openqxd

chaoos committed:
fix compiler error

chaoos committed:
use correct build flag

weinbe2 commented:
I'm happy with this PR from a visual standpoint. I still need to do build/run tests, in particular with QCD+QED, but that's more to make sure nothing unexpectedly broke, there are a few things I'd like to try beyond what's covered by our CI. This is awesome work, @chaoos .

Based on an offline convo, I'll take on updating the Wilson dslash tests to properly test recon-13 and -9 in a separate PR. This would be done by generating random U(3) instead of SU(3) fields. Since I've had to deal with this relatively recently with HISQ (as is relevant for long links) this should be relatively straightforward for me.

@maddyscientist , I believe you wanted to do some further review.

maddyscientist reviewed (approved):
Thank you @chaoos for all the work you have done and going through the review process to get this in shape to merge.

I have one small comment (regarding enabling strict prototyping for the C interface test), but beyond that, this is good to go.  Approving.

weinbe2 commented:
> Thank you @chaoos for all the work you have done and going through the review process to get this in shape to merge.
> 
> I have one small comment (regarding enabling strict prototyping for the C interface test), but beyond that, this is good to go. Approving.

I'll take point on addressing the strict prototypic for the C interface as part of updating the unit tests to explicitly check recon-9/13 when QCD+QED is enabled, I'm happy to merge this. Thanks @chaoos for all of your work (and patience when I was slow to review it)!
